File: openviking/storage/internal_names.py
"""Shared internal file-name constants for Python storage code."""

from __future__ import annotations

MULTIWRITE_PATH_LOCK_FILE = ".path.ovlock"
MULTIWRITE_EXACT_LOCK_FILE_PREFIX = ".exact.ovlock."
MULTIWRITE_REDIRECT_FILE = ".redirect.json"
MULTIWRITE_SYNC_LOG_FILE = ".sync_log.json"

MULTIWRITE_INTERNAL_FILE_NAMES = frozenset(
    {
        MULTIWRITE_PATH_LOCK_FILE,
        MULTIWRITE_REDIRECT_FILE,
        MULTIWRITE_SYNC_LOG_FILE,
    }
)


def is_storage_internal_name(name: str) -> bool:
    """Return whether ``name`` is multi-write metadata owned by the storage layer.

    Mirrors ``is_hidden_internal_name`` in ``crates/ragfs/src/core/internal_names.rs``:
    the directory lock, exact (sidecar) locks, redirect and sync-log files that RAGFS
    keeps next to user content in any directory. Such an entry is hidden from every
    listing and must never be created, written, copied or moved by a user.

    The account-root internal directories (``/local/{account}/_system`` and
    ``/local/{account}/tasks``) are deliberately not part of this predicate: root
    listings use ``VikingURI.LISTABLE_SCOPES`` as a whitelist, and below the root a
    user directory that happens to be called ``tasks`` or ``_system`` is ordinary
    content.
    """
    return name in MULTIWRITE_INTERNAL_FILE_NAMES or name.startswith(
        MULTIWRITE_EXACT_LOCK_FILE_PREFIX
    )


WEBDAV_RESERVED_FILENAMES = frozenset(
    {
        ".abstract.md",
        ".overview.md",
        ".relations.json",
        *MULTIWRITE_INTERNAL_FILE_NAMES,
    }
)

File: crates/ragfs/src/core/internal_names.rs
//! Static built-in internal names shared across the RAGFS stack.
//!
//! Every module that references `.path.ovlock`, `.exact.ovlock.*`, `.redirect.json`,
//! or `.sync_log.json` must import these constants from this single source of truth.
//! There is no dynamic extension mechanism.

/// Path-lock marker for directory-level locks.
pub const PATH_LOCK_FILE: &str = ".path.ovlock";

/// Prefix for exact (sidecar) lock files: `.exact.ovlock.<safe_name>.<sha1-prefix>`.
pub const EXACT_LOCK_FILE_PREFIX: &str = ".exact.ovlock.";

/// Multi-write redirect metadata file.
pub const REDIRECT_FILE: &str = ".redirect.json";

/// Multi-write sync-log metadata file.
pub const SYNC_LOG_FILE: &str = ".sync_log.json";

/// Returns `true` when `name` is a runtime path-lock file (`.path.ovlock` or `.exact.ovlock.*`).
pub fn is_hidden_runtime_lock_name(name: &str) -> bool {
    // Reserve the whole prefix, including names not produced by the lock resolver.
    name == PATH_LOCK_FILE || name.starts_with(EXACT_LOCK_FILE_PREFIX)
}

/// Returns `true` when `name` is any hidden internal name (lock files, redirect, sync-log).
pub fn is_hidden_internal_name(name: &str) -> bool {
    is_hidden_runtime_lock_name(name) || name == REDIRECT_FILE || name == SYNC_LOG_FILE
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn reserves_entire_exact_lock_prefix() {
        for name in [
            ".exact.ovlock.",
            ".exact.ovlock.probe.md",
            ".exact.ovlock.probe.md.0123456789abcdef",
        ] {
            assert!(is_hidden_runtime_lock_name(name));
            assert!(is_hidden_internal_name(name));
        }
        for name in [
            "probe.md",
            "tasks",
            "_system",
            ".exact.ovlock",
            "x.exact.ovlock.foo",
        ] {
            assert!(!is_hidden_internal_name(name));
        }
    }
}
